### DISCUSSION TITLE
USB HOST IN FIRMWARE set for HS 480 MbS mode?

### DISCUSSION THREAD

**Discussion by GW8RDI** [df8oe/UHSDR#1950] [General]
I was looking through the USB host code to see if I could connect a keyboard, and the initialisation routine sets the USB for HS, high speed, 480 Mb/s, which I believe requires a PHY chip to work.  Is there such a chip on the board?

Here is the code:-

`/**
  * @brief  USBH_LL_Init 
  *         Initialize the Low Level portion of the Host driver.
  * @param  phost: Host handle
  * @retval USBH Status
  */
USBH_StatusTypeDef  USBH_LL_Init (USBH_HandleTypeDef *phost)
{
  /* Init USB_IP */
  if (phost->id == HOST_HS) {		/// CAN UHSDR BE HS 480 MbS CAPABLE!?!?
  /* Link The driver to the stack */
  hhcd_USB_OTG_HS.pData = phost;
  phost->pData = &hhcd_USB_OTG_HS;

  hhcd_USB_OTG_HS.Instance = USB_OTG_HS;
  hhcd_USB_OTG_HS.Init.Host_channels = 12;
  hhcd_USB_OTG_HS.Init.speed = HCD_SPEED_FULL;
  hhcd_USB_OTG_HS.Init.dma_enable = DISABLE;
  hhcd_USB_OTG_HS.Init.phy_itface = USB_OTG_EMBEDDED_PHY;
  hhcd_USB_OTG_HS.Init.Sof_enable = DISABLE;
  hhcd_USB_OTG_HS.Init.low_power_enable = DISABLE;
  hhcd_USB_OTG_HS.Init.vbus_sensing_enable = DISABLE;
  hhcd_USB_OTG_HS.Init.use_external_vbus = DISABLE;
  if (HAL_HCD_Init(&hhcd_USB_OTG_HS) != HAL_OK)
  {
    Error_Handler();
  }

  USBH_LL_SetTimer (phost, HAL_HCD_GetCurrentFrame(&hhcd_USB_OTG_HS));
  }
  return USBH_OK;
}`

**Comment by db4ple** [comment]
This code works. It makes use of the embedded PHY of the MCU in full speed mode.  

73, DB4PLE

**Reply by GW8RDI** [reply]
Thanks, but I'm confused that the UHSDR code sets High Speed mode:-
hhcd_USB_OTG_HS.Instance = USB_OTG_HS;

The terminology is confusing, "full-speed" is "full" to 12 Mb/s, and "High speed" is up to 480 Mb/s and requires an external dedicated chip.  It's interesting that ST support have told me that for the H743 it needs the external device, or perhaps it's recommended.  I've been having some trouble getting clarity out of ST, they have confirmed that there USB drivers haven't been updated since 2015, so no support still for higher streaming rates like UAC 2.0.
